docs/build-system.org

```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
+
+The first Zig implementation [[https://github.com/candrewlee14/1brc-zig][1brc-zig]]
+| Command                                        | Mean [s]       | Min [s] | Max [s] | Relative |
+|------------------------------------------------+----------------+---------+---------+----------|
+| =./1brc-zig ~/Documents/1brc/measurements.txt= | 15.647 ± 2.206 |  13.301 |  21.649 |     1.00 |
+
+Sometimes when we redo the benchmark we don't get this huge difference.
+| Command                                        | Mean [s]       | Min [s] | Max [s] | Relative |
+|------------------------------------------------+----------------+---------+---------+----------|
+| =./1brc-zig ~/Documents/1brc/measurements.txt= | 12.749 ± 0.345 |  12.202 |  13.545 |     1.00 |
+
+
+The second Zig implementation [[https://github.com/almmiko/1brc-zig][1brc-zig - second implementation]]
+| Command                                        | Mean [s]       | Min [s] | Max [s] | Relative |
+|------------------------------------------------+----------------+---------+---------+----------|
+| =./1brc-zig ~/Documents/1brc/measurements.txt= | 18.730 ± 2.837 |  15.192 |  24.906 |     1.00 |
+
+And the C implementation
+| Command                                       | Mean [s]      | Min [s] | Max [s] | Relative |
+|-----------------------------------------------+---------------+---------+---------+----------|
+| =./analyze ~/Documents/1brc/measurements.txt= | 8.898 ± 0.591 |   7.507 |   9.510 |     1.00 |
+
+
+In the C implementation, we note that his writer uses a custom hashmap to hash the city names.
+He uses a fast multiplication hash combined with a linear search probing.
+The load factor of his hashmap is 0.5. [cite:@BillionRowChallenge]
+
+We find out that in the first Zig implementation, use also a hashmap but the Zig standard library hashmap for string hash with the wyhash [cite:@wangyiWangyifudanWyhash2024] algorithm.
+
+The ~Hotspot~ program has been used to create flamegraph of the implementations:
+
+If you look at Zig's flamegraph, you can see that 40% of the cycle time is used to insert the city into the hashmap.
+#+CAPTION: Flamegrph of the Zig's 1brc implementation
+[[../images/flamegraph_1brc_zig_1.png]]
+
+In C, on the other hand, the insertion time in the hashmap is not measurable because it is included in the ~process_chunck~ function but does not require any complex hash calculation.
+Calculating the base index is mapped into an array, and then, if there are any collisions, the index is incremented.
+#+CAPTION: Flamegrph of the C's 1brc implementation
+[[../images/flamegrah_1brc_c.png]]
+
+In the example below, this is how the hash is calculated for each city: [cite:@OneBillionRows2024a]
+#+begin_example c
+// hash everything up to ';'
+// assumption: key (city name) is at least 1 char
+unsigned int len = 1;
+unsigned int hash = (unsigned char)buf[0];
+while (buf[len] != ';') {
+  hash = (hash * 31) + (unsigned char)buf[len++];
```

Comment:
is that normal that i dont really understand what is going on here ? oupsi

Reply:
Upper, I explain more in depth the example. It's better ?